libr/io/io_map.c

```diff
@@ -9,147 +9,82 @@
 
 #define END_OF_MAP_IDS UT32_MAX
 
-#define MAP_USE_HALF_CLOSED 0
-
 #define CMP_END_GTE(addr, itv) \
 	(((addr) < r_itv_end (*(RInterval *)(itv))) ? -1 : 1)
 
-struct map_event_t {
-	RIOMap *map;
-	ut64 addr;
-	int id; // distinct priority in [0, len(maps))
-	bool is_to;
-};
-
-// Sort by address, (addr, is_to) precedes (addr, !is_to)
-static int _cmp_map_event(const void *a_, const void *b_) {
-	struct map_event_t *a = (void *)a_, *b = (void *)b_;
-	ut64 addr0 = a->addr - a->is_to, addr1 = b->addr - b->is_to;
-	if (addr0 != addr1) {
-		return addr0 < addr1? -1: 1;
-	}
-	if (a->is_to != b->is_to) {
-		return !a->is_to? -1: 1;
-	}
-	if (a->id != b->id) {
-		return a->id < b->id? -1: 1;
-	}
-	return 0;
-}
-
-static int _cmp_map_event_by_id(const void *a_, const void *b_) {
-	struct map_event_t *a = (void *)a_, *b = (void *)b_;
-	return a->id - b->id;
-}
-
-// Precondition: from == 0 && to == 0 (full address) or from < to
-static bool _map_skyline_push(RPVector *map_skyline, ut64 from, ut64 to, RIOMap *map) {
-	RIOMapSkyline *part = R_NEW (RIOMapSkyline), *part1;
-	if (!part) {
-		return false;
+#define CMP_END_GTE_PART(addr, part) \
+	(((addr) < (r_itv_end (((RIOMapSkyline *)(part))->itv)) || !r_itv_end (((RIOMapSkyline *)(part))->itv)) ? -1 : 1)
+
+#define CMP_BEGIN_GTE_PART(addr, part) \
+	(((addr) > (r_itv_begin (((RIOMapSkyline *)(part))->itv))) - ((addr) < (r_itv_begin (((RIOMapSkyline *)(part))->itv))))
+
+static bool add_map_to_skyline(RIO *io, RIOMap *map) {
+	size_t slot;
+	RPVector *skyline = &io->map_skyline;
+
+	RIOMapSkyline *new_part = R_NEW (RIOMapSkyline);
+	new_part->map = map;
+	new_part->itv = map->itv;
+	const ut64 new_part_end = r_itv_end (new_part->itv);
+
+	// `part` is the first RIOMapSkyline with part->itv.addr >= new_part->itv.addr
+	r_pvector_lower_bound (skyline, new_part->itv.addr, slot, CMP_BEGIN_GTE_PART);
+	RIOMapSkyline *part = slot < r_pvector_len (skyline) ? r_pvector_at (skyline, slot) : NULL;
+	if (slot) {
+		RIOMapSkyline *prev_part = r_pvector_at (skyline, slot - 1);
```

Comment:
Doesn't this need to check if slot is 0?

Reply:
isnt that `if (slot)` doing it?

Reply:
Sorry, not sure why I was thinking that it is `if (part)` not `if (slot)`.